### PR TITLE
centml.compile + performance prediction backend

### PR DESCRIPTION
**Pull request by yaxan**
Adds compile module with separate backend for predictions

### CODE REVIEW

File: centml/compiler/config.py [hunk 1]
```diff
@@ -31,5 +31,7 @@ class Config(BaseSettings):
     # If the server response is smaller than this, don't gzip it
     MINIMUM_GZIP_SIZE: int = 1000
 
+    PREDICTING: bool = True
```
Comment (anandj91) [inline centml/compiler/config.py:34]: instead of a boolean flag, can we use an Enum to choose different backends?

In the future, we might have more backends like compile and predict.

File: centml/compiler/prediction_sample.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import time
```
Comment (anandj91) [inline centml/compiler/prediction_sample.py:1]: these types of files shouldn't be part of the main source code. Move it either to the test/ dir or to the scripts/ dir.

File: centml/compiler/sample_data.csv
Comment (anandj91) [inline centml/compiler/sample_data.csv:1]: I don't think this file is going to be packaged with the python library. So if we generate a whl package, it won't have the sample_data.csv file.

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: centml/compiler/sample_data.csv
Comment (anandj91) [reply centml/compiler/sample_data.csv:1]: I would suggest to add a config class field to specify the location of the file for now.

File: centml/compiler/prediction_sample.py (hunk 2 above)
Comment (yaxan) [reply centml/compiler/prediction_sample.py:1]: It was going to be de deleted anyways, left it in there for initial context

File: centml/compiler/profiler.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                key = (row['op'], int(row['dim']), row['gpu'], row['dtype'])
+                points = ast.literal_eval(row['points'])
+                values = ast.literal_eval(row['values'])
+                database.add_from_db(key, points, values)
+            except ValueError as e:
+                print(f"Error parsing row: {row}")
+                print(e)
+
+
+data = './sample_data.csv'
```
[the remark this answers is not in the document]
Comment (yaxan) [reply centml/compiler/profiler.py]: How is it now?

File: centml/compile.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+from prometheus_client import start_http_server
+from torch._subclasses.fake_tensor import FakeTensorMode
+
+from centml.compiler.backend import centml_dynamo_backend
+from centml.compiler.config import OperationMode, settings
+from centml.compiler.metrics import A10_time_metric
+from centml.compiler.profiler import Profiler
+
+start_http_server(8000)
+A10_time = 0
```
Comment (anandj91) [inline centml/compile.py:14]: we should try to make the implementation a bit more generic than this. Create an environment variable to accept a list of GPU types. You can set the default value as A10 for now. Then you can write an implementation that exports metrics for all the GPU types in that list.

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: centml/compiler/profiler.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            if node.op == 'placeholder':
+                result = next(args_iter)
+            elif node.op == 'get_attr':
+                result = fetch_attr(node.target)
+            elif node.op == 'call_function':
+                args = load_arg(node.args)
+                kwargs = load_arg(node.kwargs)
+                result = node.target(*args, **kwargs)
+                inp_shapes = get_flattened_shapes(args)
+                key = (node.target.__name__, len(inp_shapes), 'A10G', 'f16')
```
Comment (anandj91) [inline centml/compiler/profiler.py:127]: get A10G from the environment variable.

get the dtype from the node.

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: centml/compile.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            backend=centml_dynamo_backend,
+            fullgraph=fullgraph,
+            dynamic=dynamic,
+            mode=mode,
+            options=options,
+            disable=disable,
+        )
+        return compiled_model
+
+    compiled_model = torch.compile(
```
Comment (anandj91) [inline centml/compile.py:55]: move this to the `else` case.

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: centml/compiler/config.py [hunk 7]
```diff
@@ -31,5 +36,10 @@ class Config(BaseSettings):
     # If the server response is smaller than this, don't gzip it
     MINIMUM_GZIP_SIZE: int = 1000
 
+    MODE: OperationMode = OperationMode.PREDICTION
+    PREDICTION_DATA_DIR: str = os.path.join(os.path.dirname(__file__), 'sample_data.csv')
```
Comment (anandj91) [inline centml/compiler/config.py:40]: lets have this variable directly point to the file instead of directory. Gives more flexibility on which filename to use.

File: centml/compile.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    *,
+    fullgraph: builtins.bool = False,
+    dynamic: Optional[builtins.bool] = None,
+    mode: Union[str, None] = None,
+    options: Optional[Dict[str, Union[str, builtins.int, builtins.bool]]] = None,
+    disable: builtins.bool = False,
+) -> Callable:
+
+    # Create a metric for each GPU
+    GPU_METRICS = {gpu: Metric(gpu) for gpu in settings.PREDICTION_GPUS.split(',')}
```
Comment (anandj91) [inline centml/compile.py:45]: We might be able to use labels instead of creating multiple Guages.

https://prometheus.github.io/client_python/instrumenting/labels/

Comment (anandj91) [reply centml/compile.py:45]: We need to check how these metrics are shown in prometheus and grafana. We already have prometheus installed in our local setup. You probably can use that. Or install prometheus and grafana by yourself.

File: centml/compile.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+from centml.compiler.config import OperationMode, settings
+from centml.compiler.profiler import Profiler
+
+start_http_server(8000)
+
+
+class Metric:
+    def __init__(self, gpu_name):
+        self._time = 0
+        self._metric = Gauge(f'{gpu_name}_metric', f'Sum of the kernel execution times on {gpu_name}')
```
Comment (anandj91) [inline centml/compile.py:18]: I feel like Histogram metric might be a better one to use. https://prometheus.github.io/client_python/instrumenting/histogram/

File: centml/compile.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+    def reset(self):
+        self._time = 0
+        self._metric.set(0)
+
+    def get_value(self):
+        return self._time
+
+
+def compile(
```
Comment (anandj91) [inline centml/compile.py:34]: We need to clean up the code a bit. Move parts of the code to relevant files,

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: centml/compile.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+            model,
+            backend=centml_dynamo_backend,  # Compilation backend
+            fullgraph=fullgraph,
+            dynamic=dynamic,
+            mode=mode,
+            options=options,
+            disable=disable,
+        )
+        return compiled_model
+    else:
```
Comment (anandj91) [inline centml/compile.py:33]: lets use else if and check if the MODE is PREDICTION.

else case should throw an except saying the MODE is not supported.

File: centml/compiler/prediction/backend.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+from prometheus_client import Gauge, start_http_server
+from torch._subclasses.fake_tensor import FakeTensorMode
+
+from centml.compiler.config import settings
+from centml.compiler.prediction.profiler import Profiler
+
+
+class GaugeMetric:
+    def __init__(self):
+        start_http_server(8000)
```
Comment (anandj91) [inline centml/compiler/prediction/backend.py:14]: this is a good place to start the server.

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: centml/compiler/prediction/profiler.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+class Profiler:
+    def __init__(self, mod, gpu):
+        self.mod = mod
+        self.graph = mod.graph
+        self.modules = dict(self.mod.named_modules())
+        self.total_time = 0
+        self.TreeDB = TreeDB()
+        self.gpu = gpu
+        populate_db(settings.PREDICTION_DATA_FILE, self.TreeDB)
```
Comment (anandj91) [inline centml/compiler/prediction/profiler.py:75]: We only need to populate the DB once right? Lets create it outside profiler and pass it in to __init__

File: centml/compiler/prediction/profiler.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+    def propagate(self, *args):
+        args_iter = iter(args)
+        env: Dict[str, Node] = {}
+
+        dtypeMap = {
```
[the remark this answers is not in the document]
Comment (yaxan) [reply centml/compiler/prediction/profiler.py:81]: yes. I had incorrectly assumed it would lead to duplicate entires (torch.float64 vs torch.double) but upon investigating they both print the same thing.

File: centml/compiler/prediction/profiler.py [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+            except ValueError as e:
+                logging.getLogger(__name__).exception(f"Error parsing row: {row}\n{e}")
+
+
+class Profiler:
+    def __init__(self, mod, gpu):
+        self.mod = mod
+        self.graph = mod.graph
+        self.modules = dict(self.mod.named_modules())
+        self.total_time = 0
```
Comment (anandj91) [inline centml/compiler/prediction/profiler.py:72]: does this need to be a class-wide variable? I feel like it can be local to the `propagate` function

Comment (anandj91) [reply centml/compiler/prediction/profiler.py:72]: As a general rule, try to define variables in their right scope. this avoids carrying over unnecessary states across the execution of different parts of the code.

File: centml/compiler/prediction/profiler.py [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+                result = fetch_attr(node.target)
+            elif node.op == 'call_function':
+                args = load_arg(node.args)
+                kwargs = load_arg(node.kwargs)
+                result = node.target(*args, **kwargs)
+
+                inp_shapes, input_dtypes = get_flattened_shapes(args)
+                output_dtypes = get_output_dtypes(result)
+
+                key = (node.target.__name__, len(inp_shapes), input_dtypes, output_dtypes, self.gpu)
```
Comment (anandj91) [inline centml/compiler/prediction/profiler.py:186]:
```suggestion
                key = (node.target, len(inp_shapes), input_dtypes, output_dtypes, self.gpu)
```

Comment (anandj91) [reply centml/compiler/prediction/profiler.py:186]: To be consistent with the rest.

File: centml/compiler/prediction/profiler.py [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+
+from centml.compiler.config import settings
+
+
+class KDTreeWithValues:
+    def __init__(self, points=None, values=None):
+        self.points = points if points else []
+        self.values = values if values else []
+        if self.points:
+            self.tree = KDTree(self.points)
```
[the remark this answers is not in the document]
Comment (yaxan) [reply centml/compiler/prediction/profiler.py:19]: It only stores coordinates, I built that class as an extension to map coordinates (input shapes) to a value (time).

File: centml/compiler/prediction/profiler.py [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+                result = node.target(*args, **kwargs)
+
+                inp_shapes, input_dtypes = get_flattened_shapes(args)
+                output_dtypes = get_output_dtypes(result)
+
+                key = (node.target.__name__, len(inp_shapes), input_dtypes, output_dtypes, self.gpu)
+
+                t = self.TreeDB.get(key, inp_shapes)
+                if t is not None:
+                    self.total_time += t
```
Comment (anandj91) [inline centml/compiler/prediction/profiler.py:190]: I would suggest the DB to always return a latency. Right now you are returning None when there is no DB entry for a key. This would mean that you are silently calculating a wrong latency.

If we cannot properly estimate the latency because of missing data, the estimated latency should somehow reflex that. One option is to return a really big latency for missing entries. Or a big negative value. Then at least we can see in the graphs that the latency calculated is wrong.

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: centml/compiler/prediction/profiler.py (hunk 16 above)
Comment (yaxan) [reply centml/compiler/prediction/profiler.py:186]: It's not a choice, sometimes node.target is a string, sometimes it's an object. In the case of a call_function it's an object. I can try cast them both to a str see if that works

File: centml/compiler/config.py [hunk 19]
```diff
@@ -9,6 +9,12 @@ class CompilationStatus(Enum):
     DONE = "DONE"
 
 
+class OperationMode(Enum):
+    PREDICTION = "PREDICTION"
+    REMOTE_COMPILATION = "REMOTE_COMPILATION"
+    TEST = "TEST"
```
Comment (anandj91) [inline centml/compiler/config.py:15]: is this just for testing?

File: centml/compiler/config.py [hunk 20]
```diff
@@ -31,5 +37,10 @@ class Config(BaseSettings):
     # If the server response is smaller than this, don't gzip it
     MINIMUM_GZIP_SIZE: int = 1000
 
+    MODE: OperationMode = OperationMode.PREDICTION
```
Comment (anandj91) [inline centml/compiler/config.py:40]: Lets use remote compilation as the default mode

File: centml/compiler/prediction/kdtree.py [hunk 21]
```diff
@@ -0,0 +1,69 @@
+import ast
+import csv
+import logging
+
+from sklearn.neighbors import KDTree  # type: ignore
+
+from centml.compiler.config import settings
+
+_tree_db = None
```
Comment (anandj91) [inline centml/compiler/prediction/kdtree.py:9]: do we need this global variable?

File: centml/compiler/prediction/metric.py [hunk 22]
```diff
@@ -0,0 +1,30 @@
+import time
+
+from prometheus_client import Gauge, start_http_server
+
+from centml.compiler.config import settings
+
+_gauge = None
```
Comment (anandj91) [inline centml/compiler/prediction/metric.py:7]: do we need this global variable?

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: centml/compiler/config.py (hunk 19 above)
Comment (yaxan) [reply centml/compiler/config.py:15]: Didn't mean to commit that

File: scripts/data_collection.py [hunk 23]
```diff
@@ -0,0 +1,355 @@
...
+        if dist > 0:
+            # print("Distance too large ", dist)
+            return None
+        return val
+
+
+db = DataCollectionTreeDB()
+
+
+class ShapeProp:
```
Comment (anandj91) [inline scripts/data_collection.py:94]: is it possible to reuse Profiler class for this? Looks like the code is almost exactly same

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: scripts/data_collection.py
Comment (anandj91) [inline scripts/data_collection.py:1]: lets make this script a bit cleaner. For instance, make it easy to configure the parameters to generate the DB file.

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: scripts/data_collection.py (hunk 23 above)
Comment (yaxan) [reply scripts/data_collection.py:94]: I could reuse some stuff but the logic is a bit different, the data collection needs to run the torch profiler every time it sees something new

File: centml/compiler/prediction/profiler.py [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+                t = self.tree_db.get(key, inp_shapes)
+
+                if self.data_collection_mode and t is None:
+                    with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CUDA]) as prof:
+                        result = node.target(*args, **kwargs)
+                    event_time_total = 0
+                    for event in prof.key_averages():
+                        event_time_total += event.cuda_time_total
+                    t = event_time_total
+                    self.tree_db.add(key, inp_shapes, t)
```
Comment (anandj91) [inline centml/compiler/prediction/profiler.py:106]: can you move this logic to a function as well? This code is repeated 3 times. You can pass in the callable operation  that need to be profiled into the function.

**Review by anandj91 — CHANGES_REQUESTED** [review]

**Review by anandj91 — APPROVED** [review]
